feat: Rename org when accepting org invite for first admin

tw4l opened the pull request:
Resolves https://github.com/webrecorder/browsertrix/issues/1874


### Manual testing
1. Log in as super user
2. Create a new org or choose an org without any users, or remove users by setting everyone to "Viewer" and then removing them
3. Update org name to be its UUID
4. Invite a new user as an admin, using an email you have access to
5. Copy-paste the invite link path in your inbox. Verify you see the "Welcome to Browsertrix" sign up form
6. Click "Next". Verify you see the org rename form with the name prefilled to `{your name}'s Archive`
7. Click "Go to Dashboard". Verify you're taken to the correct page
8. Repeat steps 2-7 for an existing user.

Regression test the following:
1. New and existing user invited after first admin
2. New and existing user invited as first admin when org name is not set to its UUID
3. Update app name and slug in org settings

### Screenshots

#### New User Invite - Step 1
<img width="956" alt="Screenshot 2024-06-25 at 4 16 47 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/7e9df551-852a-49cb-8750-baaff3cd65ce">

#### New User Invite - Step 2
<img width="983" alt="Screenshot 2024-06-25 at 4 17 03 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/fb2aa03c-b352-4bf4-ae4a-2c5447cc51bb">

#### Existing User Invite - Step 1
<img width="976" alt="Screenshot 2024-06-25 at 4 18 46 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/d9303a8c-9d8a-4b02-92d7-4d19abc0b3b8">

#### Existing User Invite - Step 2
<img width="951" alt="Screenshot 2024-06-25 at 4 41 36 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/c95b5b81-cd69-419d-850d-6630cbcedebe">

#### Taken org name error
<img width="506" alt="Screenshot 2024-06-25 at 4 41 48 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/5927a6f4-b908-4f6d-8cb1-a75499279bc8">

### Invalid invite error
<img width="490" alt="Screenshot 2024-06-25 at 5 54 49 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/01e2e85a-ae58-49b9-bd35-3c0b55b8d3e0">

tw4l committed:
Add org fields to invite endpoints

- GET /api/users/invite/token adds:
    - orgSlug
    - firstOrgAdmin (bool)
    - orgNameRequired (bool, set to true if name == oid)
- POST /api/orgs/invite-accept/token adds:
    - orgName
    - orgSlug

tw4l commented:
Oh wait, the `POST` is supposed to do a rename with those fields, missed that from the issue. Now updated.

tw4l committed:
Support renaming org via /orgs/invite-accept endpoint

SuaYoo committed:
update role select

SuaYoo committed:
handle first org admin

SuaYoo reviewed (commented):
To work with new users who need to register their password first, could we return also `orgNameRequired` in org data? That way, we can show a form to change the org name first thing after the new user is registered, and subsequent logins.

SuaYoo committed:
disable new org creation

ikreymer commented on backend/btrixcloud/users.py:
```diff
@@ -328,6 +328,17 @@ async def format_invite(self, invite):
+                result["firstOrgAdmin"] = True
+
+            # Require rename in frontend if org name is set to id
+            result["orgNameRequired"] = False
```
Should we enforce this on the backend, eg. that the orgName / orgSlug are not optional in this case?

ikreymer committed:
add type fixes

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -870,11 +876,26 @@ async def invite_user_to_org(
         invite = await invites.accept_user_invite(user, token, user_manager)
+        org = await ops.add_user_by_invite(invite, user)
+        if not org:
+            raise HTTPException(detail="invalid_invite", status_code=400)
```
Added this check, now detected with additional typing - in theory possible if non-org invite token is used, I think

SuaYoo commented:
@tw4l I updated https://github.com/webrecorder/browsertrix/issues/1874, to add optional org name and slug rename to the registration endpoint.

SuaYoo committed:
add additional fields

SuaYoo committed:
create separate org setup step

SuaYoo committed:
update accept invite

SuaYoo commented:
WIP frontend changes here https://github.com/webrecorder/browsertrix/pull/1878 set to merge into this branch

SuaYoo committed:
pass org name and slug to accept invite

SuaYoo committed:
convert invite to tailwind component

SuaYoo committed:
rename org

SuaYoo committed:
pass org id

SuaYoo committed:
handle org rename

SuaYoo committed:
handle invalid invite

SuaYoo committed:
handle accept invite

SuaYoo committed:
handle superadmin invite

SuaYoo committed:
always show org name for first admin

SuaYoo commented on backend/btrixcloud/users.py:
```diff
@@ -328,6 +328,17 @@ async def format_invite(self, invite):
         if invite.oid:
```
We'll also need org ID for the rename endpoint:
```suggestion
        if invite.oid:
            result["oid"] = invite.oid
```

SuaYoo committed:
redirect to new org

SuaYoo committed:
show steps

ikreymer committed:
lint fix

ikreymer committed:
undo optional rename in accept invite api
instead, set the org default name to '<USERNAME>'s archive' if name == id and first admin is added

tw4l committed:
Update backend/btrixcloud/users.py

Co-authored-by: sua yoo <sua@webrecorder.org>

ikreymer commented:
Should merge #1878 into this before merging.

SuaYoo committed:
update copy

SuaYoo committed:
use name returned from sign up

SuaYoo committed:
make accept two steps

SuaYoo committed:
add org description

SuaYoo committed:
adjust spacing

SuaYoo committed:
handle duplicate org name

SuaYoo committed:
Merge branch 'issue-team-7-invite-org-details' into issue-team-7-invite-org-details--frontend

SuaYoo committed:
feat: Handle first admin invite to org (#1878)

WIP frontend for https://github.com/webrecorder/browsertrix/issues/1874

### Changes

- Limits superadmin to choosing admin role when inviting the first user
to an org
- Adds org rename step after account registration
- For existing users, accepts invitation with optional org name and slug

SuaYoo committed:
handle org info returned from register

ikreymer committed:
invite accept: add "org" to api response to indicate which org user was added to

SuaYoo committed:
registration flow fixes

SuaYoo committed:
always slugify

SuaYoo committed:
clean up user info

SuaYoo committed:
show correct org info

SuaYoo committed:
update org slug after accept

SuaYoo committed:
update error

SuaYoo committed:
update user info on accept

SuaYoo committed:
remove console log

ikreymer committed:
Merge branch 'main' into issue-team-7-invite-org-details

SuaYoo committed:
update comments

ikreymer committed:
invite system cleanup:
- include sender email if not admin, include org name if not org.id, otherwise mention new org for existing users
- only set org_name if org_name != org.id for email
- typing cleanup in invites / email, make existing user email consistent with new user

tw4l commented on frontend/src/features/accounts/org-form.ts:
```diff
@@ -0,0 +1,187 @@
+        throw new Error(
+          msg("This org name or URL is already taken, try another one."),
+        );
+      }
```
Once https://github.com/webrecorder/browsertrix/pull/1891 is merged, we can also check for an invalid slug, which will return the detail `invalid_slug`.

If we want to check that an org slug is not a duplicate of an existing one ahead of time, we also have the `GET /orgs/slugs` and `GET /orgs/slug-lookup` endpoints available already in the API.

SuaYoo replied:
Moved to issue here: https://github.com/webrecorder/browsertrix/issues/1892

Shrinks99 commented on frontend/src/pages/invite/ui/inviteMessage.ts:
```diff
@@ -0,0 +1,101 @@
+
+  if (inviteInfo.firstOrgAdmin) {
+    if (isExistingUser) {
+      message = msg("Set up your organization to start web archiving.");
```
Would change to "crawling", the verb we use for running a "crawl workflow".  Also fits in with some of our other marketing copy: "Automated browser-based crawling at scale" etc.

"to start archiving" also acceptable?  We do use "archive" as a verb!

```suggestion
      message = msg("Set up your organization to start crawling.");
```

Shrinks99 commented on frontend/src/pages/invite/ui/inviteMessage.ts:
```diff
@@ -0,0 +1,101 @@
+      message = msg("Set up your organization to start web archiving.");
+    } else {
+      message = msg(
+        "Finish setting up your Browsertrix account to start web archiving.",
```
```suggestion
        "Finish setting up your Browsertrix account to start crawling.",
```

Shrinks99 commented on frontend/src/pages/invite/ui/inviteMessage.ts:
```diff
@@ -0,0 +1,101 @@
+          <h2 class="mb-3 italic text-primary">${msg("What is an org?")}</h2>
+          <p class="mb-3 text-neutral-600">
+            ${msg(
+              "An org, or organization, is a workspace for web archiving. If you’re archiving collaboratively, an org workspace can be shared between team members.",
```
Make it personal! :)

```suggestion
              "An org, or organization, is your workspace for web archiving. If you’re archiving collaboratively, an org workspace can be shared between team members.",
```

Shrinks99 commented on frontend/src/pages/invite/ui/orgAdminHelpText.ts:
```diff
@@ -0,0 +1,26 @@
+  return html`
+    <p class="mb-3 text-neutral-600">
+      ${msg(
+        "An org, or organization, is a workspace for web archiving. If you’re archiving collaboratively, an org workspace can be shared between team members.",
```
```suggestion
        "An org, or organization, is your workspace for web archiving. If you’re archiving collaboratively, an org workspace can be shared between team members.",
```

Shrinks99 commented on frontend/src/pages/org/index.ts:
```diff
@@ -197,7 +193,22 @@ export class Org extends LiteElement {
+          // TODO check if hosted instance and show support email if so
+          this.notify({
+            message: msg(
+              "You must belong to at least one org in order to log in. Please contact your Browsertrix admin to resolve the issue.",
```
How do I contact them???  Could / should we link the support email here?  Perhaps a toast is not the best place for this :\

SuaYoo replied:
This should be an edge case, and per the TODO comment if this becomes a more common issue, we can work on this once the frontend is able to differentiate the hosted instance.

Shrinks99 commented on frontend/src/pages/invite/join.ts:
```diff
@@ -0,0 +1,165 @@
+    } else if (resp.status === 404) {
+      throw new Error(
+        msg(
+          "This invite doesn't exist or has expired. Please ask the organization administrator to resend an invitation.",
```
Could we give the user some instructions on how to do this?  Support email?

SuaYoo replied:
Similar to [previous comment](https://github.com/webrecorder/browsertrix/pull/1870#discussion_r1655279776), this retains the current functionality--it looks like a new change because I renamed the file--we can re-address once we can differentiate the hosted instance.

SuaYoo committed:
Apply suggestions from code review

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

SuaYoo committed:
remove unused file

tw4l commented:
I can't approve because I started this PR but it's testing well for me, nice work!

ikreymer committed:
email: show 'Welcome to Browsertrix!' only for new users

emma-sg commented on frontend/src/pages/invite/ui/inviteMessage.ts:
```diff
@@ -0,0 +1,101 @@
+        <ul class="mt-6 text-base text-neutral-700">
+          <li class="mb-3 flex items-center gap-2">
+            <sl-icon
+              class="text-lg text-primary"
```
Can we make the logo (and maybe also the text) green when a step has been completed?
```suggestion
              class=${clsx(
                "text-lg",
                isOrgMember ? "text-success" : "text-primary",
              )}
```

emma-sg commented on frontend/src/pages/invite/ui/inviteMessage.ts:
```diff
@@ -0,0 +1,101 @@
+                  rel="noopener"
+                >
+                  org settings
+                </a>
```
Removing the extra whitespaces removes the underline from the space character following the last word here
```suggestion
                >org settings</a>
```

SuaYoo committed:
fix redirects and add tests

SuaYoo committed:
update invite message

SuaYoo committed:
fix linting warnings

ikreymer committed:
email template: only include support email line if support_email is defined

ikreymer committed:
typo fix

emma-sg commented on chart/email-templates/invite:
```diff
@@ -38,7 +44,13 @@ How do I <a href="https://docs.browsertrix.com/user-guide/archived-items/#crawl-
+
+{% if support_email %}
+<br/>
+If you're having trouble accessing the service, or if dedicated support is included in your plan, you can also contact us at: <a href="mailto:{{ support_email }}">{{ support_email }}</a>
```
```suggestion
If you're having trouble accessing the service, or if dedicated support is included in your plan, you can also contact us at <a href="mailto:{{ support_email }}">{{ support_email }}</a>
```

emma-sg commented on chart/email-templates/invite:
```diff
@@ -76,7 +96,11 @@ How do I crawl my first website? (https://docs.browsertrix.com/user-guide/workfl
+If you need any assistance, please direct your questions to the Browsertrix GitHub repo (https://github.com/webrecorder/browsertrix) or the Webrecorder community forum (https://forum.webrecorder.net).
+
+{% if support_email %}
+If you're having trouble accessing the service, or if dedicated support is included in your plan, you can also contact us at: {{ support_email }}.
```
```suggestion
If you're having trouble accessing the service, or if dedicated support is included in your plan, you can also contact us at {{ support_email }}.
```

SuaYoo committed:
update tests

SuaYoo committed:
remove unused event listener

SuaYoo committed:
oops revert unrelated changes from test branch

emma-sg reviewed (commented):
Couple tiny email suggestions -- need to test everything else more thoroughly

ikreymer committed:
Update chart/email-templates/invite

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

ikreymer committed:
Update chart/email-templates/invite

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

ikreymer reviewed (approved):
Tested fairly extensively, the sign-up / login workflows work as expected, including for both new and existing users, and renaming of org slug / org name.

I'm sure we'll have additional tweaks re: email templates, etc.. but I think this should be good!

emma-sg commented:
👍 Sounds good
